ostd/src/arch/riscv/irq/chip/plic.rs

```diff
@@ -0,0 +1,351 @@
+
+                    plic_node
+                        .property("interrupts-extended")
+                        .expect("Failed to read 'interrupts-extended' property from PLIC node")
+                        .value
+                        .chunks_exact(8)
+                        .enumerate()
+                        .filter_map(|(idx, chunk)| {
+                            let target = [
+                                u32::from_be_bytes(chunk[0..4].try_into().unwrap()),
+                                u32::from_be_bytes(chunk[4..8].try_into().unwrap()),
+                            ];
+
+                            if target[1] != 0x09 {
+                                return None;
+                            }
+
+                            let hart_id = cpu_nodes.iter().find_map(|node| {
+                                node.children()
+                                    .find(|child| {
+                                        child
+                                            .compatible()
+                                            .is_some_and(|c| c.all().any(|s| s == "riscv,cpu-intc"))
+                                            && child.property("phandle").is_some_and(|ph| {
+                                                ph.as_usize().unwrap() as u32 == target[0]
+                                            })
+                                    })
+                                    .and_then(|_| node.property("reg").and_then(|p| p.as_usize()))
+                            })?;
+
+                            Some((hart_id as u32, idx as u32))
+                        })
+                        .collect()
+                };
+
+                Self {
+                    phandle,
+                    io_mem: io_mem_builder.reserve(range),
+                    num_interrupt_sources,
+                    hart_to_target_mapping,
+                    interrupt_number_mappings: (0..num_interrupt_sources)
+                        .map(|_| None)
+                        .collect::<Vec<Option<u32>>>()
+                        .into_boxed_slice(),
+                }
+            })
+            .collect()
+    }
+
+    pub(super) fn map_interrupt_source_to(
+        &mut self,
+        interrupt_source: u32,
+        irq_line: &IrqLine,
+    ) -> Result<()> {
+        assert!(interrupt_source > 0 && interrupt_source < self.num_interrupt_sources);
+        match self.interrupt_number_mappings[interrupt_source as usize] {
+            None => {
+                self.interrupt_number_mappings[interrupt_source as usize] =
+                    Some(irq_line.num() as u32);
+                Ok(())
+            }
+            Some(_) => Err(Error::InvalidArgs),
+        }
+    }
+
+    pub(super) fn unmap_interrupt_source(&mut self, interrupt_source: u32) {
+        assert!(interrupt_source > 0 && interrupt_source < self.num_interrupt_sources);
+        self.interrupt_number_mappings[interrupt_source as usize] = None;
+    }
```

Comment:
I think the `pub(super) interrupt_number_mappings()` method should be removed. We can add a method here:
```rust
    pub(super) fn interrupt_source_mapping(&self, interrupt_source: u32) -> Option<u32>;
```
It seems better with the methods `unmap_interrupt_source`/`map_interrupt_source`.

Reply:
https://github.com/asterinas/asterinas/pull/2106/commits/f657cea371145fe3ed88db43f22a7fd84cdb20de#r2443299186 also means that we should move `interrupt_number_mappings`  to the second `impl Plic` block because it's very related to `unmap_interrupt_source`/`map_interrupt_source` methods.

Do you have any particular reason for leaving it in the first `impl` block? If so, you can ignore this comment. I'm also okay with that.